src/Microsoft.OData.Core/JsonLight/ODataJsonLightOutputContext.cs

```diff
@@ -939,29 +886,37 @@ protected override async ValueTask DisposeAsyncCore()
         /// Creates a new JSON writer of <see cref="IJsonWriter"/>.
         /// </summary>
         /// <param name="container">The dependency injection container to get related services.</param>
-        /// <param name="textWriter">The text writer to write to.</param>
+        /// <param name="stream">The stream to write to.</param>
         /// <param name="isIeee754Compatible">true if the writer should write large integers as strings.</param>
         /// <param name="writerSettings">Configuration settings for the OData writer.</param>
+        /// <param name="encoding">The character encoding to use.</param>
         /// <returns>An <see cref="IJsonWriter"/> instance.</returns>
         private static IJsonWriter CreateJsonWriter(
             IServiceProvider container,
-            TextWriter textWriter,
+            Stream stream,
             bool isIeee754Compatible,
-            ODataMessageWriterSettings writerSettings)
+            ODataMessageWriterSettings writerSettings,
+            Encoding encoding)
         {
             IJsonWriter jsonWriter;
             if (container == null)
             {
+                TextWriter textWriter = new StreamWriter(stream, encoding);
+
                 jsonWriter = new JsonWriter(textWriter, isIeee754Compatible);
             }
             else
             {
                 IJsonWriterFactory jsonWriterFactory = container.GetRequiredService<IJsonWriterFactory>();
-                jsonWriter = jsonWriterFactory.CreateJsonWriter(textWriter, isIeee754Compatible);
-                Debug.Assert(jsonWriter != null, "jsonWriter != null");
+                jsonWriter = jsonWriterFactory.CreateJsonWriter(stream, isIeee754Compatible, encoding);
+
+                if (jsonWriter == null)
+                {
+                    throw new ODataException(Strings.ODataMessageWriter_JsonWriterFactory_ReturnedNull(isIeee754Compatible, encoding.WebName));
+                }
             }
 
-            if (jsonWriter is Json.JsonWriter defaultJsonWriter)
+            if (jsonWriter is JsonWriter defaultJsonWriter)
             {
                 defaultJsonWriter.ArrayPool = writerSettings.ArrayPool;
             }

```

Comment:
nit: This block is a code smell in my opinion. Seems like the json writer factory abstraction isn't that clean if we have to resort to these checks. But don't know if there's a quick way to fix without investigating time in some design changes (which is likely outside the scope). But I do hope we get rid of this array pool one day.

Maybe the `ArrayPool` could be a property of the `DefaultJsonWriterFactory` instead, and it could pass it down to the `JsonWriter`. Feel free to ignore this comment if it distracts from the scope.

Reply:
@habbes I created an issue to track the redesign for a cleaner approach https://github.com/OData/odata.net/issues/2881